File: code/15. 3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        if (n < 3) return {};
        sort(begin(nums), end(nums));
        vector<vector<int>> ans;
        int a = 1e5 + 5;
        for (int i = 0; i < n - 2; i++) {
            if (a == nums[i]) continue;
            a = nums[i];
            int left = i + 1, right = n - 1;
            while (left < right) {
                if (nums[left] + nums[right] + a > 0) {
                    right--;
                } else if (nums[left] + nums[right] + a < 0) {
                    left++;
                } else {
                    ans.push_back({a, nums[left], nums[right]});
                    left++;
                    right--;
                    while (left < right && nums[left] == nums[left - 1]) left++;
                }
            }
        }
        return ans;
    }
};

File: code/15. 3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        if n < 3:
            return list()
        nums.sort()
        a = 1e5 + 5
        ans = list()
        for i in range(0, n - 2):
            if a == nums[i]:
                continue
            a = nums[i]
            left = i + 1
            right = n - 1
            while left < right:
                if a + nums[left] + nums[right] > 0:
                    right -= 1
                elif a + nums[left] + nums[right] < 0:
                    left += 1
                else:
                    ans.append([a, nums[left], nums[right]])
                    left += 1
                    right -= 1
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
        return ans
        